File: src/Java/myservicebus/src/main/java/com/myservicebus/persistence/OutboxMessage.java
package com.myservicebus.persistence;

import java.net.URI;
import java.time.Instant;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.UUID;

public record OutboxMessage(
        UUID recordId,
        UUID messageId,
        OutboxDeliveryIntent intent,
        URI destinationAddress,
        List<String> messageTypes,
        byte[] body,
        String contentType,
        Map<String, String> headers,
        Instant createdAtUtc,
        UUID requestId,
        UUID correlationId,
        UUID conversationId,
        UUID initiatorId,
        URI responseAddress,
        URI faultAddress,
        Instant availableAtUtc) {

    public OutboxMessage(
            UUID recordId,
            UUID messageId,
            OutboxDeliveryIntent intent,
            URI destinationAddress,
            List<String> messageTypes,
            byte[] body,
            String contentType,
            Map<String, String> headers,
            Instant createdAtUtc,
            UUID requestId,
            UUID correlationId,
            UUID conversationId,
            UUID initiatorId,
            URI responseAddress,
            URI faultAddress) {
        this(recordId, messageId, intent, destinationAddress, messageTypes, body, contentType, headers,
                createdAtUtc, requestId, correlationId, conversationId, initiatorId, responseAddress, faultAddress,
                createdAtUtc);
    }

    public OutboxMessage {
        requireNonEmpty(recordId, "recordId");
        requireNonEmpty(messageId, "messageId");
        Objects.requireNonNull(intent, "intent");
        Objects.requireNonNull(destinationAddress, "destinationAddress");
        Objects.requireNonNull(messageTypes, "messageTypes");
        Objects.requireNonNull(body, "body");
        Objects.requireNonNull(headers, "headers");
        Objects.requireNonNull(createdAtUtc, "createdAtUtc");
        Objects.requireNonNull(availableAtUtc, "availableAtUtc");
        if (messageTypes.isEmpty() || messageTypes.stream().anyMatch(value -> value == null || value.isBlank())) {
            throw new IllegalArgumentException("At least one non-empty message type is required.");
        }
        if (contentType == null || contentType.isBlank()) {
            throw new IllegalArgumentException("contentType must not be blank.");
        }
        messageTypes = List.copyOf(messageTypes);
        body = body.clone();
        headers = Map.copyOf(headers);
    }

    @Override
    public byte[] body() {
        return body.clone();
    }

    private static void requireNonEmpty(UUID value, String name) {
        Objects.requireNonNull(value, name);
        if (value.equals(new UUID(0, 0))) {
            throw new IllegalArgumentException(name + " must not be empty.");
        }
    }
}

File: src/MyServiceBus/Persistence/OutboxMessage.cs
namespace MyServiceBus.Persistence;

using System.Collections.ObjectModel;

public enum OutboxDeliveryIntent
{
    Send,
    Publish,
    Reply,
    Fault
}

public enum OutboxMessageState
{
    Pending,
    Leased,
    Dispatched,
    Dead,
    Cancelled
}

public sealed class OutboxMessage
{
    private readonly byte[] body;

    public OutboxMessage(
        Guid recordId,
        Guid messageId,
        OutboxDeliveryIntent intent,
        Uri destinationAddress,
        IReadOnlyList<string> messageTypes,
        ReadOnlySpan<byte> body,
        string contentType,
        IReadOnlyDictionary<string, string> headers,
        DateTimeOffset createdAtUtc,
        Guid? requestId = null,
        Guid? correlationId = null,
        Guid? conversationId = null,
        Guid? initiatorId = null,
        Uri? responseAddress = null,
        Uri? faultAddress = null,
        DateTimeOffset? availableAtUtc = null,
        DateTimeOffset? scheduledAtUtc = null,
        Guid? causationMessageId = null)
    {
        ArgumentOutOfRangeException.ThrowIfEqual(recordId, Guid.Empty);
        ArgumentOutOfRangeException.ThrowIfEqual(messageId, Guid.Empty);
        ArgumentNullException.ThrowIfNull(destinationAddress);
        ArgumentNullException.ThrowIfNull(messageTypes);
        ArgumentException.ThrowIfNullOrWhiteSpace(contentType);
        ArgumentNullException.ThrowIfNull(headers);

        if (messageTypes.Count == 0 || messageTypes.Any(string.IsNullOrWhiteSpace))
            throw new ArgumentException("At least one non-empty message type is required.", nameof(messageTypes));

        RecordId = recordId;
        MessageId = messageId;
        Intent = intent;
        DestinationAddress = destinationAddress;
        MessageTypes = Array.AsReadOnly(messageTypes.ToArray());
        this.body = body.ToArray();
        ContentType = contentType;
        Headers = new ReadOnlyDictionary<string, string>(new Dictionary<string, string>(headers));
        CreatedAtUtc = createdAtUtc;
        RequestId = requestId;
        CorrelationId = correlationId;
        ConversationId = conversationId;
        InitiatorId = initiatorId;
        ResponseAddress = responseAddress;
        FaultAddress = faultAddress;
        AvailableAtUtc = availableAtUtc ?? createdAtUtc;
        ScheduledAtUtc = scheduledAtUtc;
        CausationMessageId = causationMessageId;
    }

    public Guid RecordId { get; }
    public Guid MessageId { get; }
    public OutboxDeliveryIntent Intent { get; }
    public Uri DestinationAddress { get; }
    public IReadOnlyList<string> MessageTypes { get; }
    public ReadOnlyMemory<byte> Body => body.ToArray();
    public string ContentType { get; }
    public IReadOnlyDictionary<string, string> Headers { get; }
    public DateTimeOffset CreatedAtUtc { get; }
    public Guid? RequestId { get; }
    public Guid? CorrelationId { get; }
    public Guid? ConversationId { get; }
    public Guid? InitiatorId { get; }
    public Uri? ResponseAddress { get; }
    public Uri? FaultAddress { get; }
    public DateTimeOffset AvailableAtUtc { get; }
    public DateTimeOffset? ScheduledAtUtc { get; }
    public Guid? CausationMessageId { get; }
}
